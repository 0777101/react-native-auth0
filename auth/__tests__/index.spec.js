import Auth from '../';
import fetchMock from 'fetch-mock';

describe('auth', () => {

  const baseUrl = 'samples.auth0.com';
  const clientId = 'A_CLIENT_ID_OF_YOUR_ACCOUNT';
  const telemetry = {name: 'react-native-auth0', version: '1.0.0'};
  const redirectUri = 'https://mysite.com/callback';
  const state = 'a random state for auth';
  const tokens = {
    status: 200,
    body: {
      access_token: 'an access token',
      id_token: 'an id token',
      expires_in: 1234567890,
      state,
      scope: 'openid'
    },
    headers: { 'Content-Type': 'application/json' }
  };
  const oauthError = {
    status: 400,
    body: {
      error: 'invalid_request',
      error_description: 'Invalid grant'
    },
    headers: { 'Content-Type': 'application/json' }
  };
  const unexpectedError = {
    status: 500,
    body: 'Internal Server Error....',
    headers: { 'Content-Type': 'text/plain' }
  };
  const auth  = new Auth({baseUrl, clientId, telemetry});

  beforeEach(fetchMock.restore);

  describe('constructor', () => {
    it('should build with domain', () => {
      const auth = new Auth({baseUrl, clientId});
      expect(auth.clientId).toEqual(clientId);
    });

    it('should fail without clientId', () => {
      expect(() => new Auth({baseUrl})).toThrowErrorMatchingSnapshot();
    });

    it('should fail without domain', () => {
      expect(() => new Auth({clientId})).toThrowErrorMatchingSnapshot();
    });
  });


  describe('authorizeUrl', () => {
    it('should return default authorize url', () => {
      expect(auth.authorizeUrl({
        responseType: 'code',
        redirectUri,
        state: 'a_random_state'
      })).toMatchSnapshot();
    });

    it('should return default authorize url with extra parameters', () => {
      expect(auth.authorizeUrl({
        responseType: 'code',
        redirectUri,
        state: 'a_random_state',
        connection: 'facebook'
      })).toMatchSnapshot();
    });
  });

  describe('code exchange', () => {
    it('should send correct payload', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', tokens);
      expect.assertions(1);
      await auth.exchange({code: 'a code', verifier: 'a verifier', redirectUri, state, scope: 'openid'});
      expect(fetchMock.lastCall()).toMatchSnapshot();
    });

    it('should return successful response', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', tokens);
      expect.assertions(1);
      const parameters = {code: 'a code', verifier: 'a verifier', redirectUri, state, scope: 'openid'};
      await expect(auth.exchange(parameters)).resolves.toMatchSnapshot();
    });

    it('should handle oauth error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', oauthError);
      expect.assertions(1);
      const parameters = {code: 'a code', verifier: 'a verifier', redirectUri, state, scope: 'openid'};
      await expect(auth.exchange(parameters)).rejects.toMatchSnapshot();
    });

    it('should handle unexpected error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', unexpectedError);
      expect.assertions(1);
      const parameters = {code: 'a code', verifier: 'a verifier', redirectUri, state, scope: 'openid'};
      await expect(auth.exchange(parameters)).rejects.toMatchSnapshot();
    });
  });

  describe('password realm', () => {
    const parameters = {username: 'info@auth0.com', password: 'secret pass', realm: 'Username-Password-Authentication', audience: 'http://myapi.com', scope: 'openid'};
    it('should send correct payload', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', tokens);
      expect.assertions(1);
      await auth.realm(parameters);
      expect(fetchMock.lastCall()).toMatchSnapshot();
    });

    it('should return successful response', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', tokens);
      expect.assertions(1);
      await expect(auth.realm(parameters)).resolves.toMatchSnapshot();
    });

    it('should handle oauth error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', oauthError);
      expect.assertions(1);
      await expect(auth.realm(parameters)).rejects.toMatchSnapshot();
    });

    it('should handle unexpected error', async () => {
      fetchMock.postOnce('https://samples.auth0.com/oauth/token', unexpectedError);
      expect.assertions(1);
      await expect(auth.realm(parameters)).rejects.toMatchSnapshot();
    });
  });

});