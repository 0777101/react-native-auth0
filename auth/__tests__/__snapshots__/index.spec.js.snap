// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth authorizeUrl should return default authorize url 1`] = `"https://samples.auth0.com/authorize?response_type=code&redirect_uri=https%3A%2F%2Fmysite.com%2Fcallback&state=a_random_state&client_id=A_CLIENT_ID_OF_YOUR_ACCOUNT&auth0Client=eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0"`;

exports[`auth authorizeUrl should return default authorize url with extra parameters 1`] = `"https://samples.auth0.com/authorize?response_type=code&redirect_uri=https%3A%2F%2Fmysite.com%2Fcallback&state=a_random_state&connection=facebook&client_id=A_CLIENT_ID_OF_YOUR_ACCOUNT&auth0Client=eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0"`;

exports[`auth code exchange should handle oauth error 1`] = `[invalid_request: Invalid grant]`;

exports[`auth code exchange should handle unexpected error 1`] = `[a0.response.invalid: Internal Server Error]`;

exports[`auth code exchange should return successful response 1`] = `
Object {
  "accessToken": "an access token",
  "expiresIn": 1234567890,
  "idToken": "an id token",
  "scope": "openid",
  "state": "a random state for auth",
}
`;

exports[`auth code exchange should send correct payload 1`] = `
Array [
  "https://samples.auth0.com/oauth/token",
  Object {
    "body": "{\\"code\\":\\"a code\\",\\"code_verifier\\":\\"a verifier\\",\\"redirect_uri\\":\\"https://mysite.com/callback\\",\\"client_id\\":\\"A_CLIENT_ID_OF_YOUR_ACCOUNT\\",\\"grant_type\\":\\"authorization_code\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`auth constructor should fail without clientId 1`] = `"Missing clientId in parameters"`;

exports[`auth constructor should fail without domain 1`] = `"Missing Auth0 domain"`;

exports[`auth password realm should handle oauth error 1`] = `[invalid_request: Invalid grant]`;

exports[`auth password realm should handle unexpected error 1`] = `[a0.response.invalid: Internal Server Error]`;

exports[`auth password realm should return successful response 1`] = `
Object {
  "accessToken": "an access token",
  "expiresIn": 1234567890,
  "idToken": "an id token",
  "scope": "openid",
  "state": "a random state for auth",
}
`;

exports[`auth password realm should send correct payload 1`] = `
Array [
  "https://samples.auth0.com/oauth/token",
  Object {
    "body": "{\\"username\\":\\"info@auth0.com\\",\\"password\\":\\"secret pass\\",\\"realm\\":\\"Username-Password-Authentication\\",\\"audience\\":\\"http://myapi.com\\",\\"scope\\":\\"openid\\",\\"client_id\\":\\"A_CLIENT_ID_OF_YOUR_ACCOUNT\\",\\"grant_type\\":\\"http://auth0.com/oauth/grant-type/password-realm\\"}",
    "headers": Object {
      "Accept": "application/json",
      "Auth0-Client": "eyJuYW1lIjoicmVhY3QtbmF0aXZlLWF1dGgwIiwidmVyc2lvbiI6IjEuMC4wIn0",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;
